File: android/app/src/main/java/com/mitimaiti/app/viewmodels/OnboardingViewModel.kt
package com.mitimaiti.app.viewmodels

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mitimaiti.app.models.Gender
import com.mitimaiti.app.models.Intent
import com.mitimaiti.app.models.ShowMe
import com.mitimaiti.app.services.APIService
import com.mitimaiti.app.services.PhotoRepository
import com.mitimaiti.app.services.UserPrefs
import com.mitimaiti.app.utils.ImageCompression
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

enum class OnboardingStep { NAME, BIRTHDAY, GENDER, PHOTOS, INTENT, SHOW_ME, LOCATION, READY }

class OnboardingViewModel : ViewModel() {
    private val _currentStep = MutableStateFlow(OnboardingStep.NAME)
    val currentStep: StateFlow<OnboardingStep> = _currentStep.asStateFlow()
    val firstName = MutableStateFlow(UserPrefs.firstName.value); val isNonSindhi = MutableStateFlow(false)

    init {
        // Persist the name to the shared store whenever it changes, so
        // ProfileViewModel can show the right name after onboarding.
        viewModelScope.launch {
            firstName.collect { UserPrefs.setFirstName(it) }
        }
    }
    val birthDay = MutableStateFlow(""); val birthMonth = MutableStateFlow(""); val birthYear = MutableStateFlow("")
    val selectedGender = MutableStateFlow<Gender?>(null)
    val selectedPhotos: StateFlow<List<Uri>> = PhotoRepository.photos
    val selectedIntent = MutableStateFlow<Intent?>(null); val selectedShowMe = MutableStateFlow<ShowMe?>(null)
    val selectedCity = MutableStateFlow("")

    val age: Int? get() {
        val y = birthYear.value.toIntOrNull() ?: return null; val m = birthMonth.value.toIntOrNull() ?: return null; val d = birthDay.value.toIntOrNull() ?: return null
        return try { java.time.Period.between(java.time.LocalDate.of(y, m, d), java.time.LocalDate.now()).years } catch (e: Exception) { null }
    }
    val isAgeValid: Boolean get() = (age ?: 0) >= 18
    val progress: Float get() = (OnboardingStep.entries.indexOf(_currentStep.value) + 1).toFloat() / OnboardingStep.entries.size
    val canProceed: Boolean get() = when (_currentStep.value) {
        OnboardingStep.NAME -> firstName.value.isNotBlank(); OnboardingStep.BIRTHDAY -> isAgeValid
        OnboardingStep.GENDER -> selectedGender.value != null; OnboardingStep.PHOTOS -> selectedPhotos.value.isNotEmpty()
        OnboardingStep.INTENT -> selectedIntent.value != null; OnboardingStep.SHOW_ME -> selectedShowMe.value != null
        OnboardingStep.LOCATION -> selectedCity.value.isNotBlank(); OnboardingStep.READY -> true
    }

    fun nextStep() { val s = OnboardingStep.entries; val i = s.indexOf(_currentStep.value); if (i < s.size - 1) _currentStep.value = s[i + 1] }
    fun previousStep() { val s = OnboardingStep.entries; val i = s.indexOf(_currentStep.value); if (i > 0) _currentStep.value = s[i - 1] }
    fun addPhoto(uri: Uri) { PhotoRepository.addPhoto(uri) }
    fun removePhoto(index: Int) { PhotoRepository.removePhoto(index) }

    private val _isSubmitting = MutableStateFlow(false)
    val isSubmitting: StateFlow<Boolean> = _isSubmitting.asStateFlow()
    private val _submitError = MutableStateFlow<String?>(null)
    val submitError: StateFlow<String?> = _submitError.asStateFlow()

    fun submitOnboarding(context: Context, onDone: (Boolean) -> Unit) {
        android.util.Log.d("Onboarding", "submitOnboarding called, isSubmitting=${_isSubmitting.value}")
        if (_isSubmitting.value) { android.util.Log.w("Onboarding", "early return: already submitting"); return }
        viewModelScope.launch {
            android.util.Log.d("Onboarding", "coroutine started")
            _isSubmitting.value = true
            _submitError.value = null

            val basics = mutableMapOf<String, Any>()
            if (firstName.value.isNotBlank()) basics["display_name"] = firstName.value.trim()
            val y = birthYear.value.toIntOrNull(); val m = birthMonth.value.toIntOrNull(); val d = birthDay.value.toIntOrNull()
            if (y != null && m != null && d != null) {
                basics["date_of_birth"] = "%04d-%02d-%02d".format(y, m, d)
            }
            selectedGender.value?.let { g ->
                basics["gender"] = when (g) {
                    Gender.MAN -> "man"; Gender.WOMAN -> "woman"; Gender.NON_BINARY -> "non-binary"
                }
            }
            if (selectedCity.value.isNotBlank()) basics["city"] = selectedCity.value.trim()

            val userFields = mutableMapOf<String, Any>()
            selectedIntent.value?.let { i ->
                userFields["intent"] = when (i) {
                    Intent.CASUAL -> "casual"; Intent.OPEN -> "open"
                    Intent.SERIOUS, Intent.MARRIAGE -> "marriage"
                }
            }

            val settings = mutableMapOf<String, Any>()
            selectedShowMe.value?.let { s ->
                settings["gender_preference"] = when (s) {
                    ShowMe.MEN -> "men"; ShowMe.WOMEN -> "women"; ShowMe.EVERYONE -> "everyone"
                }
            }

            val payload = mutableMapOf<String, Any>()
            if (basics.isNotEmpty()) payload["basics"] = basics
            if (userFields.isNotEmpty()) payload["user"] = userFields
            if (settings.isNotEmpty()) payload["settings"] = settings

            android.util.Log.d("Onboarding", "calling updateProfile with payload=$payload")
            val profileResult = APIService.updateProfile(payload)
            android.util.Log.d("Onboarding", "updateProfile result: isSuccess=${profileResult.isSuccess}, exceptionOrNull=${profileResult.exceptionOrNull()}")
            if (profileResult.isFailure) {
                _submitError.value = "Couldn't save profile. Please try again."
                _isSubmitting.value = false
                onDone(false)
                return@launch
            }

            for (uri in selectedPhotos.value) {
                val bytes = ImageCompression.compressForUpload(context, uri) ?: continue
                APIService.uploadPhoto(bytes)
            }

            _isSubmitting.value = false
            onDone(true)
        }
    }
}

File: ios/MitiMaiti/ViewModels/OnboardingViewModel.swift
import SwiftUI
import PhotosUI

@MainActor
class OnboardingViewModel: ObservableObject {
    @Published var currentStep: OnboardingStep = .name
    @Published var firstName = UserProfileStore.shared.firstName {
        didSet { UserProfileStore.shared.firstName = firstName }
    }
    @Published var isNonSindhi = false
    @Published var birthDay = 15
    @Published var birthMonth = 6
    @Published var birthYear = 1998
    @Published var selectedGender: Gender?
    @Published var selectedPhotos: [String] = []
    @Published var selectedImages: [UIImage] = []
    @Published var selectedIntent: Intent?
    @Published var selectedShowMe: ShowMe?
    @Published var selectedCity: String?
    @Published var selectedRegion: String?
    @Published var selectedCountry: String?
    @Published var isLoading = false
    @Published var error: String?

    // Tracks which selectedImages indices have already been uploaded so going
    // back-and-forth between steps doesn't re-upload the same photo.
    private var uploadedImageHashes: Set<Int> = []
    private let api = APIService.shared

    init() {
        // No local name (e.g. phone-OTP signup) — the server may still know one
        // (backfilled from a Google/Apple link). Fetch and prefill, but never
        // clobber anything the user has started typing.
        if firstName.isEmpty {
            Task { [weak self] in
                guard let user = try? await APIService.shared.fetchProfile() else { return }
                guard let self, self.firstName.isEmpty, !user.displayName.isEmpty else { return }
                self.firstName = user.displayName
            }
        }
    }

    var age: Int {
        let components = DateComponents(year: birthYear, month: birthMonth, day: birthDay)
        guard let dob = Calendar.current.date(from: components) else { return 0 }
        return Calendar.current.dateComponents([.year], from: dob, to: Date()).year ?? 0
    }

    var isAgeValid: Bool {
        age >= 18
    }

    var progress: Double {
        Double(currentStep.rawValue + 1) / Double(OnboardingStep.allCases.count)
    }

    var canProceed: Bool {
        switch currentStep {
        case .name: return firstName.count >= 2
        case .birthday: return isAgeValid
        case .gender: return selectedGender != nil
        case .photos: return selectedImages.count >= 1
        case .intent: return selectedIntent != nil
        case .showMe: return selectedShowMe != nil
        case .location: return selectedCity != nil
        case .ready: return true
        }
    }

    func nextStep() {
        guard canProceed else { return }
        if let next = OnboardingStep(rawValue: currentStep.rawValue + 1) {
            withAnimation(.spring(response: 0.4, dampingFraction: 0.8)) {
                currentStep = next
            }
        }
    }

    func previousStep() {
        if let prev = OnboardingStep(rawValue: currentStep.rawValue - 1) {
            withAnimation(.spring(response: 0.4, dampingFraction: 0.8)) {
                currentStep = prev
            }
        }
    }

    func addPhoto() {
        if selectedPhotos.count < 6 {
            selectedPhotos.append("photo_\(selectedPhotos.count + 1)")
        }
    }

    func addImage(_ image: UIImage) {
        guard selectedImages.count < 6 else { return }
        let newIndex = selectedImages.count
        selectedImages.append(image)
        selectedPhotos.append("photo_\(selectedPhotos.count + 1)")
        // Persist every photo; index 0 is the primary/avatar
        UserImageStore.shared.save(image, at: newIndex)
    }

    func removePhoto(at index: Int) {
        guard index < selectedPhotos.count else { return }
        selectedPhotos.remove(at: index)
        if index < selectedImages.count {
            selectedImages.remove(at: index)
        }
        // Keep the store in sync with the current selection
        UserImageStore.shared.setAll(selectedImages)
    }

    /// Upload every photo in selectedImages that we haven't already pushed
    /// to the backend, then advance to the next step. Called from the
    /// Continue button on the photos step so the .ready / Discover screens
    /// see the photos on the user's actual profile.
    func proceedFromPhotos() async {
        guard canProceed else { return }
        isLoading = true
        error = nil

        // Reconcile with the server first. Re-running onboarding or resuming
        // must not blindly append past the 6-photo cap — find how many photos
        // the account already has and only upload enough net-new ones to fill.
        let maxPhotos = 6
        var serverPhotoCount = 0
        if let profile = try? await api.fetchProfile() {
            serverPhotoCount = profile.photos.count
        }
        var remainingSlots = max(0, maxPhotos - serverPhotoCount)
        var uploadedThisPass = false

        for (index, image) in selectedImages.enumerated() {
            let hash = image.hashValue
            if uploadedImageHashes.contains(hash) { continue }
            // Account already at capacity (e.g. photos from a prior run) —
            // nothing more to upload, the requirement is already met.
            if remainingSlots <= 0 { break }
            guard let data = image.jpegData(compressionQuality: 0.85) else {
                continue
            }
            do {
                _ = try await api.uploadPhoto(imageData: data)
                uploadedImageHashes.insert(hash)
                remainingSlots -= 1
                uploadedThisPass = true
            } catch APIError.photoLimitReached {
                // Server says we're already at the limit — not an onboarding
                // failure; the user has enough photos. Stop trying to add more.
                break
            } catch {
                self.error =
                    "Couldn't upload photo \(index + 1): \(error.localizedDescription)"
                isLoading = false
                return
            }
        }

        // Onboarding only needs the account to end up with at least one photo.
        if serverPhotoCount == 0 && !uploadedThisPass && uploadedImageHashes.isEmpty {
            self.error = "Please add at least one photo to continue."
            isLoading = false
            return
        }

        isLoading = false
        nextStep()
    }

    /// Persist every onboarding-collected field to the backend in one
    /// PATCH /me. Photos are uploaded earlier (proceedFromPhotos); this
    /// covers the rest, which until now never left the device. Called once
    /// from the final screen. Returns the backend-recalculated completeness,
    /// or nil if nothing to send / the request failed.
    func submitProfile() async -> Int? {
        var basics: [String: Any] = [:]
        let name = firstName.trimmingCharacters(in: .whitespaces)
        if name.count >= 2 { basics["display_name"] = name }
        basics["date_of_birth"] = String(
            format: "%04d-%02d-%02d", birthYear, birthMonth, birthDay
        )
        if let g = selectedGender { basics["gender"] = g.rawValue }
        if let city = selectedCity?.trimmingCharacters(in: .whitespaces),
           !city.isEmpty {
            basics["city"] = city
        }
        if let st = selectedRegion?.trimmingCharacters(in: .whitespaces),
           !st.isEmpty {
            basics["state"] = st
        }
        if let co = selectedCountry?.trimmingCharacters(in: .whitespaces),
           !co.isEmpty {
            basics["country"] = co
        }

        var userFields: [String: Any] = [:]
        if let intent = selectedIntent { userFields["intent"] = intent.rawValue }

        var settings: [String: Any] = [:]
        if let show = selectedShowMe {
            settings["gender_preference"] = show.rawValue
        }

        var payload: [String: Any] = [:]
        if !basics.isEmpty { payload["basics"] = basics }
        if !userFields.isEmpty { payload["user"] = userFields }
        if !settings.isEmpty { payload["settings"] = settings }
        guard !payload.isEmpty else { return nil }

        do {
            return try await api.patchMe(payload)
        } catch {
            self.error = "Couldn't save your profile: \(error.localizedDescription)"
            return nil
        }
    }
}
